drake/core/Function.h

```diff
@@ -104,46 +104,24 @@ namespace Drake {
     }
   };
 
-  class Function {
-  public:
-    Function() : relation(InputOutputRelation::Form::ARBITRARY) {};
-    Function(InputOutputRelation::Form f) : relation(f) {};
-    virtual ~Function() {};
-
-    virtual void eval(const Eigen::Ref<const Eigen::VectorXd>& x, Eigen::VectorXd& y) const = 0;
-    virtual const InputOutputRelation& getInputOutputRelation() const { return relation; };
-    // note: for multiple argument functions, this can be getInputOutputRelation(input_num,output_num)
-
-  protected:
-    InputOutputRelation relation;
-  };
-
-  class DifferentiableFunction : public Function {
-  public:
-    DifferentiableFunction() : Function(InputOutputRelation::Form::DIFFERENTIABLE) {};
-    virtual ~DifferentiableFunction() {};
-
-    virtual void eval(const Eigen::Ref<const Eigen::VectorXd>& x, Eigen::VectorXd& y) const = 0;
-    virtual void eval(const Eigen::Ref<const TaylorVecXd>& x, TaylorVecXd& y) const = 0;
-  };
-
-  template <typename Derived>
-  class TemplatedDifferentiableFunction : public DifferentiableFunction {
-  public:
-    TemplatedDifferentiableFunction(Derived& derived) : DifferentiableFunction(), derived(derived) {};
-    virtual ~TemplatedDifferentiableFunction() {};
+  template <typename ScalarType> using VecIn = Eigen::Ref< Eigen::Matrix<ScalarType, Eigen::Dynamic, 1> const >;
+  template <typename ScalarType> using VecOut = Eigen::Matrix<ScalarType, Eigen::Dynamic, 1>;
 
-    virtual void eval(const Eigen::Ref<const Eigen::VectorXd>& x, Eigen::VectorXd& y) const override { derived.evalImpl(x,y); }
-    virtual void eval(const Eigen::Ref<const TaylorVecXd>& x, TaylorVecXd& y) const override { derived.evalImpl(x,y); }
-
-  private:
-    Derived& derived;
+  /** FunctionTraits
+   * @brief Define interface to a function of the form y = f(x).
+   */
+  template <typename F>
+  struct FunctionTraits {
+    // TODO: add in/out relation, possibly distinguish differentiable functions
+    static size_t numInputs(F const& f) { return f.numInputs(); }
```

Comment:
assuming this is what you mean, can we call this numInputArguments instead?  (we use numInputs to ask about the dimension of the vector in, e.g. the System concept)